api/orders.js

```diff
@@ -281,8 +281,223 @@ function cancelOrder(request, response, next) {
+    return new Promise(function(resolve, reject) {
+      if (!ripple.UInt160.is_valid(options.account)) {
+        reject(new InvalidRequestError('Parameter is not a valid Ripple address: account'));
+      }
+
+      if (!options.base.currency) {
+        reject(new InvalidRequestError('Invalid parameter: base. Must be a currency string in the form currency+counterparty'));
+      }
+
+      if (!validator.isValid(options.base.currency, 'Currency')) {
+        reject(new InvalidRequestError('Invalid parameter: base. Must be a currency string in the form currency+counterparty'));
+      }
+
+      if (options.base.currency !== 'XRP' && (!options.base.issuer || !ripple.UInt160.is_valid(options.base.issuer))) {
+        reject(new InvalidRequestError('Invalid parameter: base. Must be a currency string in the form currency+counterparty'));
+      }
+
+      if (!options.counter.currency) {
+        reject(new InvalidRequestError('Invalid parameter: counter. Must be a currency string in the form currency+counterparty'));
+      }
+
+      if (!validator.isValid(options.counter.currency, 'Currency')) {
+        reject(new InvalidRequestError('Invalid parameter: counter. Must be a currency string in the form currency+counterparty'));
+      }
+
+      if (options.counter.currency !== 'XRP' && (!options.counter.issuer || !ripple.UInt160.is_valid(options.counter.issuer))) {
+        reject(new InvalidRequestError('Invalid parameter: counter. Must be a currency string in the form currency+counterparty'));
+      }
+
+      if (options.counter.currency === 'XRP' && options.counter.issuer) {
+        reject(new InvalidRequestError('Invalid parameter: counter. XRP cannot have counterparty'));
+      }
+
+      if (options.base.currency === 'XRP' && options.base.issuer) {
+        reject(new InvalidRequestError('Invalid parameter: base. XRP cannot have counterparty'));
+      }
+
+      resolve(options);
+    });
+  }
+
+  function getLastValidatedLedger(options) {
+    var promise = new Promise(function (resolve, reject) {
+      var ledgerRequest = remote.requestLedger('validated');
+
+      ledgerRequest.once('success', function(res) {
+        options.ledger = res.ledger.ledger_index;
+        resolve(options);
+      });
+
+      ledgerRequest.once('error', reject);
+      ledgerRequest.request();
+    });
+
+    return promise;
+  }
+
+  function getBookOffers(taker_gets, taker_pays, options) {
+    var promise = new Promise(function (resolve, reject) {
+      var bookOffersRequest = remote.requestBookOffers({
+        gets: taker_gets,
+        pays: taker_pays,
+        ledger: options.ledger,
+        limit: options.limit,
+        taker: options.account
+      });
+
+      bookOffersRequest.once('success', resolve);
+      bookOffersRequest.once('error', reject);
+      bookOffersRequest.request();
+    });
+
+    return promise;
+  }
+
+  function getBids(options) {
+    var taker_gets = options.counter;
+    var taker_pays = options.base;
```

Comment:
These are the names used by rippled. And by `amount` you mean `value` right? I wanted to keep the variable naming consistent with the response object, since it is constructed from these intermediates.